sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/thetasketchesExpressions.scala

```diff
@@ -0,0 +1,266 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.datasketches.common.SketchesArgumentException
+import org.apache.datasketches.memory.Memory
+import org.apache.datasketches.theta.{CompactSketch, SetOperation}
+
+import org.apache.spark.sql.catalyst.expressions.{ExpectsInputTypes, Expression, ExpressionDescription, Literal}
+import org.apache.spark.sql.catalyst.expressions.codegen.CodegenFallback
+import org.apache.spark.sql.catalyst.util.ThetaSketchUtils
+import org.apache.spark.sql.errors.QueryExecutionErrors
+import org.apache.spark.sql.types.{AbstractDataType, BinaryType, DataType, IntegerType, LongType}
+
+@ExpressionDescription(
+  usage = """
+    _FUNC_(expr) - Returns the estimated number of unique values given the binary representation
+    of a Datasketches ThetaSketch. """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_(theta_sketch_agg(col)) FROM VALUES (1), (1), (2), (2), (3) tab(col);
+       3
+  """,
+  group = "misc_funcs",
+  since = "4.1.0")
+case class ThetaSketchEstimate(child: Expression)
+    extends UnaryExpression
+    with CodegenFallback
+    with ExpectsInputTypes {
+  override def nullIntolerant: Boolean = true
+
+  override protected def withNewChildInternal(newChild: Expression): ThetaSketchEstimate =
+    copy(child = newChild)
+
+  override def prettyName: String = "theta_sketch_estimate"
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(BinaryType)
+
+  override def dataType: DataType = LongType
+
+  override def nullSafeEval(input: Any): Any = {
+    val buffer = input.asInstanceOf[Array[Byte]]
+    try {
+      Math.round(CompactSketch.wrap(Memory.wrap(buffer)).getEstimate)
+    } catch {
+      case _: SketchesArgumentException | _: java.lang.Error =>
```

Comment:
The `java.lang.Error` is very broad and could cover many types of different problems that could arise from the above math operations. To improve the health of the code, would you mind to convert it to one or more specific exception types instead, and ideally split the operations on L59 above into multiple steps and surround only the appropriate step with the applicable exceptions of interest? This may result in multiple try/catch blocks. Same below for other functions in this file.

Reply:
I'll look into this. I agree the` java.lang.Error` is broad.